File: (3) unsafe_example/main.cpp
#include <cstdint>
#include <vector>


struct Foo {
    int bar;
    double baz;
};


class FooContainer {
public:
    FooContainer(size_t num_elements) { 
        buffer.resize(sizeof(Foo) * num_elements);
    }

    Foo& Get(size_t index) {
        return reinterpret_cast<Foo&>(buffer[sizeof(Foo) + index]);
    }

private:
    std::vector<uint8_t> buffer;
};


int main() {
    auto container = FooContainer(1024);
    auto& first = container.Get(0);
}

File: (3) unsafe_example/src/main.rs
#![allow(unused)]

use std::mem::{size_of, transmute};

struct Foo {
    bar: u32,
    baz: f64,
}

struct FooContainer {
    buffer: Vec<u8>,
}

impl FooContainer {
    pub fn new(num_elements: usize) -> Self {
        let mut buffer = Vec::new();
        buffer.resize(size_of::<Foo>() * num_elements, 0);

        FooContainer { buffer }
    }

    pub unsafe fn get_unsafe(&mut self, index: usize) -> &Foo {
        transmute(&mut self.buffer[size_of::<Foo>() * index])
    }

    pub fn get_safe(&mut self, index: usize) -> &Foo {
        unsafe {
            return transmute(&mut self.buffer[size_of::<Foo>() * index]);
        }
    }
}

fn main() {
    let mut container = FooContainer::new(1024);

    unsafe {
        let first = container.get_unsafe(0);
    };

    let first = container.get_safe(0);
}
